Make Subscriptor thread-safe & add general list_subscribers() function.

jppelteret opened the pull request:
Fixes #6314

drwells commented:
I'm having some deja-vu here (from ec871a9a585a34cabc831c33cff4201d62a35ef7) :)

I'll look this over later today. I want to be sure that we don't have some large penalty to invoke for creating things inheriting from `Subscriptor` like `Vector`.

jppelteret commented:
Hmm... I see, so that's what `atomic` means... Would the right way to do this to remove the locks in `subscribe` and `unsubscribe`, while leaving the one in list_subscribers` to prevent interleaving of the output stream?

For the record, I'm not completely opposed to just chucking out this PR if you think its going to cause performance issues (although these locks are only used in debug mode, and bypassed altogether in release mode). That said, the changes I made here helped me solve my problem, so being able to print a list of subscribers regardless of the MT configuration could potentially be helpful to someone else in the future.

tamiko commented on source/base/subscriptor.cc:
```diff
@@ -185,15 +188,15 @@ Subscriptor::unsubscribe(const char *id) const
-#  ifndef DEAL_II_WITH_THREADS
+#ifdef DEAL_II_WITH_THREADS
+  std::lock_guard<std::mutex> lock(mutex);
+#endif
```
I am not sure whether we actually want to serialize with a lock_guard here. Even though this is only relevant for the debug mode - I fear that a mutex here might prevent a whole class of race conditions showing up in debug mode and only in release mode. Making debugging actually harder.

drwells replied:
I agree with @tamiko here: we should have the same behavior in both modes.

drwells commented on include/deal.II/base/subscriptor.h:
```diff
@@ -236,6 +250,24 @@ Subscriptor::serialize(Archive &,
+        << counter << " subscriptions from \""
+        << it->first << '\"'
+        << std::endl;
+}
```
I don't think that using a lock is necessary here. I believe that it should be up to the caller to not print to object from multiple threads simultaneously.

masterleinad replied:
I think we do. We still want to make sure that the map doesn't change when running through all elements of the map.
We could use a `shared_lock` here as well.

drwells replied:
Thats a good point. We should keep the lock for this reason.

jppelteret replied:
> We could use a shared_lock here as well.

@masterleinad Apparently this is a [c++14 feature](http://en.cppreference.com/w/cpp/thread/shared_lock), so we'll have to do with the mutex + lock guard option.

masterleinad replied:
I know, but we could `boost::shared_lock`. However, I am not sure if this is worth it.

jppelteret replied:
Ah right, I forgot that such functionality would be covered by boost. Thanks :-)

jppelteret commented:
Thanks for all of the comments! I can't address them right now, but I'll take a shot at making the appropriate changes tomorrow.

drwells commented:
If we update this then we need to also further update the explanatory text in step-6.

jppelteret commented:
@drwells Either way, the documentation in step-6 should be updated. Towards the end ("From the above error message...") the output error message relates to a latent subscription to a `SphericalManifold`, but the discussion refers to a subscription to an `FE_Q`. I'll check what the new output is and touch up the documentation here. 

Was this the only place that you anticipated modifications to be necessary? The discussion in the destructor section (`Step6::~Step6`) looks OK to me.

drwells commented:
Nope, see #6381. My current PR is up to date with the current implementation.

drwells commented:
I think the only other place where we would have to update step-6 would be to mention that this behavior is always on in debug mode (the current text is deliberately ambiguous here).

jppelteret commented:
/run-tests

davydden commented:
@drwells @jppelteret are we merging this in 9.0?

jppelteret commented:
I can see that I'd have to fix a couple of tests before it could be merged. I'm not sure its that important to get this in 9.0. What do you think @drwells?

drwells commented:
I would prefer not to add any non bug fixes (or documentation changes, e.g., step-60) to 9.0 now.

jppelteret commented:
Agreed. I've set the milestone for this to the next release.

drwells commented:
Thanks. There is so much stuff to do that we really do need to start excluding things now, unfortunately.

jppelteret commented:
@drwells I have rebased and fixed the tests now. I have also added a changelog entry. Would you mind please checking if you are happy with the state of the PR?

drwells commented on doc/news/changes/minor/20180715Jean-PaulPelteret:
```diff
@@ -0,0 +1,6 @@
+Changed: The Subscriptor class has been made thread-safe, and its behavior in
+debug and release modes are now aligned. Although this makes subscription and 
```
I think it would make more sense to write "its behavior in debug and release modes are now identical."

drwells commented on source/base/subscriptor.cc:
```diff
@@ -151,14 +156,14 @@ Subscriptor::operator=(Subscriptor &&s) noexcept
-#  ifndef DEAL_II_WITH_THREADS
+#ifdef DEAL_II_WITH_THREADS
+  std::lock_guard<std::mutex> lock(mutex);
+#endif
```
I think this should go above the assignment to `object_info` in case two threads try to write to that pointer at the same time. Is there a reason that it is below that I am missing?

Alternatively, we could make `object_info` atomic.

jppelteret replied:
I think that you might be right. I've fixed this now.

drwells commented:
ping @jppelteret :)

jppelteret commented:
@drwells Thanks for the reminder. This 5 minutes worth of work should have been completed at the hackathon. I feel ashamed... Nevertheless, I've now attended to your comments.

drwells commented:
There is no rush on this :) I am just glad that it is done. We have way too many open PRs right now!
